src/extras/reload.js

```diff
@@ -0,0 +1,69 @@
+(function (global) {
+  global.TRACING = true;
+
+  const systemJSPrototype = System.constructor.prototype;
+  const dependentsForModule = {};
+
+  systemJSPrototype.reload = function (id, parentUrl) {
+    const loader = this;
+
+    // An import first sets up a promise and resolves the module id, recreating
+    // the same behavior here avoids a race condition if a reload is called within
+    // that period
+    return Promise.resolve()
+      .then(function () {
+        return loader.resolve(id, parentUrl);
+      })
+      .then(function (id) {
+        const oldLoad = loader.getLoad(id);
+        if (!oldLoad) {
+          // module was not loaded before
+          return loader.import(id);
+        }
+
+        // module was imported before, and it might still be in a loading state
+        // first do a regular import for this module to ensure a current executing load
+        // is finished. except on error, as that will throw the same error again
+        const init = !oldLoad.er ? loader.import(id) : Promise.resolve();
```

Comment:
The need to check this internal state can be avoided by always waiting on the `loader.import` of the id being hot-reloaded.

Reply:
I tried that, but I was running into the issue I mentioned above. I'll see if I can get a reproduction of that.

Reply:
Ok let me know then we can either fix the bug or decide what to do further.

Reply:
Actually, this line isn't to fix the situation I mentioned above. This was to enable being able to recover from an error during the previous import. But we can do this without an internal state check by adding an empty catch handler.